### ISSUE TITLE
Setting Seed on R possibly not working

### ISSUE THREAD

**Issue by blewy** [rstudio/tensorflow#114]
I'm reviewing some code that compares the Sigmoid activation function and RElu. 
I've set a seed for r and  tensorflow session, but the results change every time I redo the experiment. Its seems it's not working for replication purposes.

The code is very simple:

Sys.setenv(TENSORFLOW_PYTHON="/anaconda/bin/python")
library(tensorflow)
  # Create graph
  sess = tf$Session()
  
  #Setting the seeds
  tf$set_random_seed(seed = 5)
  set.seed(42)
  
  batch_size = 50
  iterations=500
  learning_rate = 0.01 
  
  a1 = tf$Variable(tf$random_normal(shape(1L,1L), mean = 0,stddev = 1))
  b1 = tf$Variable(tf$random_uniform(shape(1L,1L),minval = 0,maxval = 1))
  a2 = tf$Variable(tf$random_normal(shape(1L,1L), mean = 0,stddev = 1))
  b2 = tf$Variable(tf$random_uniform(shape(1L,1L),minval = 0, maxval = 1))
                   
  x = rnorm(500, mean = 2, sd = 0.1)
  x_data = tf$placeholder(tf$float32, list(NULL,1L))
  
  #Activation functions: Sigmoid
  sigmoid_activation = tf$sigmoid(tf$add(tf$matmul(x_data, a1), b1))
  
  #Activation functions: Relu
  relu_activation = tf$nn$relu(tf$add(tf$matmul(x_data, a2), b2))
  
  # Declare the loss function as the difference between
  # the output and a target value, 0.75.
  loss1 = tf$reduce_mean(tf$square(tf$subtract(sigmoid_activation, 0.75)))
  loss2 = tf$reduce_mean(tf$square(tf$subtract(relu_activation, 0.75)))
  
  # Initialize variables
  init = tf$global_variables_initializer()
  sess$run(init)
  
  # Declare optimizer
  my_opt = tf$train$GradientDescentOptimizer(learning_rate)
  train_step_sigmoid = my_opt$minimize(loss1)
  train_step_relu = my_opt$minimize(loss2)
  
  loss_vec_sigmoid = c() 
  loss_vec_relu = c() 
  for (i in 1:iterations){
  rand_index = as.vector(sample.int(length(x), size=batch_size,replace = FALSE, prob = NULL))
  rand_x = as.matrix(x[rand_index])
  
  sess$run(train_step_sigmoid, feed_dict=dict(x_data= rand_x))
  sess$run(train_step_relu, feed_dict=dict(x_data= rand_x))
  #save loss
  loss_vec_sigmoid<- c(loss_vec_sigmoid,sess$run(loss1, feed_dict=dict(x_data=rand_x)))
  loss_vec_relu<- c(loss_vec_relu,sess$run(loss2, feed_dict=dict(x_data=rand_x)))
  #save 
  sigmoid_output<-mean(sess$run(sigmoid_activation, feed_dict=dict(x_data=rand_x)))
  relu_output<- mean(sess$run(relu_activation,feed_dict=dict(x_data=rand_x)))
  
  if ((i+1)%%50==0){
    cat('Step #' , (i+1),"\n" ) # ,' A = ', sess$run(A) , ' b = ' , (sess$run(b)),"\n" )
    cat('sigmoid = ',mean(loss_vec_sigmoid),' relu = ', mean(loss_vec_relu),"\n")
  }
  
  }
  sess$close()

**Comment by terrytangyuan** [comment]
I noticed something similar recently as well. Seems like you did it correctly by setting graph level seed. Could you try setting seed for individual ops, e.g. set `seed` in `tf$random_normal`, etc.?

**Comment by terrytangyuan** [comment]
Are you using nightly builds of TensorFlow?

**Comment by traversc** [comment]
Hi Terry, is there any effective difference between using TensorFlow functions rather than the equivalent R functions for variable initialization?

**Comment by terrytangyuan** [comment]
@traversc Not too much statistical difference if other contingencies are met, e.g. seed, etc. TensorFlow functions initialize those variables in terms of Tensors. There's a lot of benefits computationally, e.g. gradient updates, optimizations, distributed variable placements/persistence, etc.

**Comment by terrytangyuan** [comment]
@traversc Well, I meant no statistical difference if their data types are the same. TensorFlow provides a lot of dtypes that you can use in variable initializers that may make a lot of statistical difference, especially in fields that require different precisions.

**Comment by blewy** [comment]
hi terrytangyuan, didnt catch your sujection. you want me to test if setting the seed inside every variable inicialization works ?

**Comment by terrytangyuan** [comment]
Yes try that. Also try to see if it's reproducible in Python. You maybe
able to find some answers on StackOverflow useful as well.

**Comment by blewy** [comment]
yep...work just fine.

**Comment by terrytangyuan** [comment]
@jjallaire anything you can think of that might mess up the seed?

**Comment by jjallaire** [comment]
If I restart R between runs I get the exact same results on every run:

```r
Step # 50 
sigmoid =  0.3609613  relu =  0.1737563 
Step # 100 
sigmoid =  0.327267  relu =  0.08630343 
Step # 150 
sigmoid =  0.2879187  relu =  0.05754694 
Step # 200 
sigmoid =  0.2477292  relu =  0.0432306 
Step # 250 
sigmoid =  0.2121787  relu =  0.03467915 
Step # 300 
sigmoid =  0.1830558  relu =  0.02897638 
Step # 350 
sigmoid =  0.1598485  relu =  0.02491086 
Step # 400 
sigmoid =  0.1413459  relu =  0.02185962 
Step # 450 
sigmoid =  0.1264132  relu =  0.01948926 
Step # 500 
sigmoid =  0.114195  relu =  0.01759335 
```
